ecosystem/auto-relayer/auto-relayer.md

```diff
@@ -0,0 +1,210 @@
+- Cost Efficient: the service must optimize both operational costs and users fees to ensure economic viability for operators and affordability for users.
+
+## Proposed Solution
+
+### Architecture diagram
+
+![Architecture Diagram](architecture_diagram.png)
+
+This architecture will be repeated for each interop dependency set (devnet, sepolia, and mainnet). This is a multi-chain architecture grouped by dependency set and each component will handle traffic across all chains within the dependency set. As the interop dependency set expands we may start to run into scaling bottlenecks, from there we can determine which services to split out into dedicated instances per chain.
+
+### Database schema
+
+#### Sent Messages
+db: `event-log-indexer-postgres`
+| name | type   |
+|------|--------|
+| messageHash   | string | 
+| sourceTxHash   | string | 
+| chainID   | bigint | 
+| destinationChainID   | bigint | 
+| target   | string | 
+| messageOrigin   | string | 
+| sender   | string | 
+| txOrigin   | string | 
+| messagePayload   | string | 
+| messageNonce   | bigint | 
+| blockNumber   | bigint | 
+| logIndex   | int | 
+| timestamp   | bigint | 
+| rootMessageHash | string |
+| depth | int |
+| status | enum |
+
+#### Relayed Messages
+db: `event-log-indexer-postgres`
+| name | type   |
+|------|--------|
+| messageHash   | string | 
+| rootMessageHash | string |
+| depth | int |
+| chainID | string |
+| txHash | string |
+| timestamp | bigint |
+| relayer | string |
+| cost | bigint |
+
+#### Gas Tank by Message Hash
+db: `event-log-indexer-postgres`
+| name | type   |
+|------|--------|
+| rootMsgHash | string |
+| balance | bigint |
+
+#### Gas Tank by Address
+db: `event-log-indexer-postgres`
+| name | type   |
+|------|--------|
+| account | address |
+| balance | bigint |
+
+#### Refunds
+db: `event-log-indexer-postgres`
+| name | type   |
+|------|--------|
+| messageHash | string |
+| messageTxOrigin | string |
+| status | enum |
+| refundTxHash | string | 
+
+#### Relay Attempts
+db: `relayer-postgres`
+| name | type   |
+|------|--------|
+| messageHash | string |
+| timestamp | bigint
+| result | enum |
+| relayer | string |
+
+### Libraries
+- [Ponder](https://github.com/ponder-sh/ponder) for indexing of onchain events.
+- [Bullmq](https://github.com/taskforcesh/bullmq) for message queue.
+- Postgres database. 
+
+### Resource Usage
+
+#### Database Resource Usage
+The most significant source of resource usage for the database will be the indexing of all `SentMessage` and `RelayMessage` events from the `L2toL2CrossDomainMessenger`. Initially we will store all of these events without a retention period, however, as resource usage increases we can make optimizations to remove messages that have expired or are no longer needed because they have been relayed.
+
+### Transaction submission
```

Comment:
Could we have Ponder stream events directly into a Redis/Bullmq queue via its event hooks instead of persisting everything first to Postgres and then polling? I think that might cut relay latency and reduce write load on the DB.

Reply:
I like that. We should add event hooks to ponder to directly put new events on the queue.  If a balance in the fee vault, can cover that message, then the message can immediately be relayed. Persisting is still useful in the case where the message cant be relayed immediately at emission (because fee vault balance is too low) and then fee is added to the fee vault post-message being sent.